archive/100-theorems-list/88_derangements_formula.lean

```diff
@@ -0,0 +1,468 @@
+    exact f_derangement a },
+
+  calc derangements α
+      ≃ Σ b : α, {f : derangements α // f a = b}
+          : (sigma_preimage_equiv _).symm
+  ... ≃ Σ b : everything_but a, {f : derangements α // f a = b}
+          : (sigma_subtype_equiv_of_subset _ _ fact).symm
+  ... ≃ Σ b : everything_but a, (derangements (neither_of a b) ⊕ derangements (everything_but a))
+          : sigma_congr_right (equiv_one_fiber a),
+end
+
+end recursion
+
+
+section finite_instances
+
+variables {α : Type*} [decidable_eq α] [fintype α]
+
+instance : decidable_pred (@is_derangement α) :=
+begin
+  intro f,
+  apply fintype.decidable_forall_fintype,
+end
+
+instance : fintype (derangements α) :=
+begin
+  have : fintype (perm α) := by apply_instance,
+  dsimp [derangements],
+  exact set_fintype (set_of is_derangement),
+end
+
+instance (a : α) : fintype (everything_but a) :=
+begin
+  simp [everything_but],
+  apply fintype.of_finset ({a}ᶜ : finset α),
+  simp,
+end
+
+instance (a : α) (b : everything_but a) : fintype (neither_of a b) :=
+begin
+  simp [everything_but, neither_of],
+  apply fintype.of_finset ({a, b}ᶜ : finset α),
+  simp,
+end
+
+end finite_instances
+
+
+section counting
+
+/-- The number of derangements on an `n`-element set. -/
+def num_derangements (n : ℕ) : ℕ := card (derangements (fin n))
+
+lemma num_derangements_invariant (α : Type*) [fintype α] [decidable_eq α] :
+  card (derangements α) = num_derangements (card α) :=
+begin
+  unfold num_derangements,
+  apply card_eq.mpr,  -- card_eq because we don't need the specific equivalence
+  use derangement_congr (equiv_fin α),
+end
+
+theorem num_derangements_recursive (n : ℕ) :
+  num_derangements (n+2) = (n+1) * num_derangements n + (n+1) * num_derangements (n+1) :=
+begin
+  let X := fin(n+2),
+  let star : X := 0, -- could be any element tbh
+
+  have card_everything_but : card (everything_but star) = (n + 1),
+  { simp [everything_but, finset.card_compl] },
+
+  have card_neither_of : ∀ m : everything_but star, card (neither_of star m) = n,
+  { simp [everything_but, finset.card_compl],
+    intros m hm,
+
+    suffices : finset.card ({star, m} : finset X) = 2,
+    { simp [this] },
+
+    -- TODO this is very clunky, what tricks can i use here?
+    rw finset.card_insert_of_not_mem,
+    { simp },
+    { change _ ≠ _ at hm,
+      simp [hm.symm] }
+  },
+
+  have card_der_1 : card (derangements (everything_but star)) = num_derangements (n+1),
+  { rw [num_derangements_invariant, card_everything_but] },
+
+  have card_der_2 : ∀ m : everything_but star, card (derangements (neither_of star m)) = num_derangements n,
+  { intro m, rw [num_derangements_invariant, card_neither_of] },
+
+  have key := card_congr (derangements_recursion_equiv star),
+  rw num_derangements_invariant X at key,
+  simp [card_der_1, card_der_2] at key,
+
+  -- TODO dunno why this didn't work out with simp, but whatever
+  simp [fintype.card] at card_everything_but,
+  rw [key, card_everything_but],
+  ring,
```

Comment:
```suggestion
  rw key,
  simpa [finset.card_univ, card_everything_but] using mul_add _ _ _,
```

Reply:
what's the difference between the `using` clause, and just putting it inside the simp?

Reply:
it basically does `simp` and then `exact`s what's in the `using` clause when it can; putting inside the `simp` could work too! i didn't do it because I worried it could loop (if something brackets everything back up again) but it probably works just fine